### PR TITLE
Compile Releases Transform - include records

### PR DESCRIPTION
**Pull request by odscjames**
https://github.com/open-contracting/kingfisher-process/issues/63

Also, in ocdskingfisherprocess/transform/compile_releases.py marked private methods as private with a _ at start

### CODE REVIEW

**Comment by odscjames** [comment]
@jpmckinney I can add more tests and there is some logging to do, but can you just confirm the order we try things in in `ocdskingfisherprocess/transform/compile_releases.py` is what you were expecting?

**Comment by odscjames** [comment]
[Process: Add support for records to compile-release transform](https://trello.com/c/ZjcZkv3C/191-process-add-support-for-records-to-compile-release-transform)

File: ocdskingfisherprocess/transform/compile_releases.py [hunk 1]
```diff
@@ -55,18 +56,105 @@ def get_ocids(self):
...
+            self._process_record(ocid, records[0])
+
+        else:
+
+            self._process_releases(ocid)
+
+    def _process_record(self, ocid, record):
+
+        releases = record.get('releases', [])
+        releases_linked = [r for r in releases if 'url' in r and r['url']]
```
Comment (jpmckinney) [inline ocdskingfisherprocess/transform/compile_releases.py:113]: This isn't a robust test for a linked release – though the OC4IDS script in OCDS Kit uses the same logic. I'll add an `ocdskit.util.is_linked_release` method to OCDS Kit.

I think there was a discussion at some point (perhaps with @Bjwebb and Duncan) about how best to identify linked releases, but I can't find it.

File: ocdskingfisherprocess/transform/compile_releases.py [hunk 2]
```diff
@@ -55,18 +56,105 @@ def get_ocids(self):
...
+
+            # This counts as taking a random record, as we have not ordered the SQL query by anything.
+            # (In practice the way postgres works I think we will always get the first record by load order,
+            #     but with no ORDER BY clause that is not guaranteed)
+            # TODO log we have done this
+            self._process_record(ocid, records[0])
+
+        elif len(records) == 1:
+
+            self._process_record(ocid, records[0])
```
Comment (jpmckinney) [inline ocdskingfisherprocess/transform/compile_releases.py:104]: Since the processing is the same, I think it's better to have:

```python
if records:
    if len(records) > 1:
        # log message
    self._process_record(ocid, records[0])
```

Same for a similar pattern with `releases_compiled` below.

**Comment by odscjames** [comment]
> This isn't a robust test for a linked release – though the OC4IDS script in OCDS Kit uses the same logic. I'll add an ocdskit.util.is_linked_release method to OCDS Kit.

@jpmckinney Let me know when this is released.

I did some more tests but wasn't sure how far to go, given the impending Django work. There are a few more edge cases to cover if you want, or this could be ok for now?

File: ocdskingfisherprocess/transform/compile_releases.py [hunk 3]
```diff
@@ -55,18 +56,112 @@ def get_ocids(self):
...
+                records.append(self.database.get_data(row['data_id']))
+
+        # Decide what to do .....
+        if len(records) > 1:
+
+            # This counts as taking a random record, as we have not ordered the SQL query by anything.
+            # (In practice the way postgres works I think we will always get the first record by load order,
+            #     but with no ORDER BY clause that is not guaranteed)
+            warning = 'There were multiple records for this OCID! ' + \
+                    'We have picked one at random and passed it through this transform.'
```
Comment (jpmckinney) [inline ocdskingfisherprocess/transform/compile_releases.py:100]: Change warning:

* were -> are
* We have picked one at random … -> The record to pass through was selected arbitrarily.

And remove the comment. Same change for similar warning below.

In general, for all messages:

* Use present tense.
* Use passive voice instead of "we".

File: ocdskingfisherprocess/transform/compile_releases.py [hunk 4]
```diff
@@ -55,18 +56,112 @@ def get_ocids(self):
...
+
+            self._process_record(ocid, records[0])
+
+        else:
+
+            self._process_releases(ocid)
+
+    def _process_record(self, ocid, record, warnings=None):
+
+        # avoid https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments
```
Comment (jpmckinney) [inline ocdskingfisherprocess/transform/compile_releases.py:113]: You can remove this comment. This is a very common pattern for default arguments.

File: ocdskingfisherprocess/transform/compile_releases.py [hunk 5]
```diff
@@ -55,18 +56,112 @@ def get_ocids(self):
...
+
+        if len(releases_compiled) > 1:
+            # If more than one, pick one at random. and log that.
+            warnings.append('This already has multiple compiled releases in the source! ' +
+                            'We have picked one at random and passed it through this transform unchanged.')
+            self._store_result(ocid, releases_compiled[0], warnings=warnings)
+
+        elif len(releases_compiled) == 1:
+            # There is just one compiled release - pass it through unchanged, and log that.
+            warnings.append('This already has one compiled release in the source! ' +
```
Comment (jpmckinney) [inline ocdskingfisherprocess/transform/compile_releases.py:146]: This message is easy to confuse with another above. Change to:

* one compiled release in the releases array

Same for similar warning above.

File: ocdskingfisherprocess/transform/compile_releases.py [hunk 6]
```diff
@@ -55,18 +56,112 @@ def get_ocids(self):
...
+    def _process_record(self, ocid, record, warnings=None):
+
+        # avoid https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments
+        if not warnings:
+            warnings = []
+
+        releases = record.get('releases', [])
+        releases_linked = [r for r in releases if 'url' in r and r['url']]
+
+        if len(releases) > 0 and len(releases_linked) == 0:
```
Comment (jpmckinney) [inline ocdskingfisherprocess/transform/compile_releases.py:120]: Just do `if releases and and not releases_linked`

File: ocdskingfisherprocess/transform/compile_releases.py [hunk 7]
```diff
@@ -55,18 +56,112 @@ def get_ocids(self):
...
+            # There is just one compiled release - pass it through unchanged, and log that.
+            warnings.append('This already has one compiled release in the source! ' +
+                            'We have passed it through this transform unchanged.')
+            self._store_result(ocid, releases_compiled[0], warnings=warnings)
+
+        else:
+            # We can't process this ocid. Warn of that.
+            self.database.add_collection_note(
+                self.destination_collection.database_id,
+                'OCID ' + ocid + ' could not be complied as we did not have enough data to do that.'
```
Comment (jpmckinney) [inline ocdskingfisherprocess/transform/compile_releases.py:154]: Typo: complied -> compiled

The warning is very unspecific. We should be clear:

* as we … -> , because at least one release in the releases array is a linked release, and the record has neither a compileRelease nor a release with a tag of "compiled".

**Comment by jpmckinney** [comment]
@odscjames OCDS Kit is released: https://ocdskit.readthedocs.io/en/latest/changelog.html

**Comment by odscjames** [comment]
To update OCDS Kit we need to update OCDS Merge, so this is blocked until https://github.com/open-contracting/kingfisher-process/pull/280 is sorted

**Comment by odscjames** [comment]
Changes made
